File: src/main/java/org/kite9/diagram/common/elements/vertex/MultiCornerVertex.java
package org.kite9.diagram.common.elements.vertex;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.commons.math.fraction.BigFraction;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.model.position.HPos;
import org.kite9.diagram.model.position.Layout;
import org.kite9.diagram.model.position.VPos;
import org.kite9.framework.common.Kite9ProcessingException;

/**
 * Represents corners of diagrams, containers and any other rectangular content.
 * For elements with grid-layout, these also represent points within the grid that will need to be connected up.
 * Multi-corners can be the corners of multiple different diagram elements.
 */
public class MultiCornerVertex extends AbstractAnchoringVertex implements MultiElementVertex {
	
	public static final boolean isMin(BigFraction b) {
		return b.equals(BigFraction.ZERO);
	}
	
	public static final boolean isMax(BigFraction b) {
		return b.equals(BigFraction.ONE);
	}
	
	
	public static BigFraction getOrdForXDirection(Direction d) {
		switch (d) {
		case LEFT:
			return BigFraction.ZERO;
		case RIGHT:
			return BigFraction.ONE;
		default:
			return BigFraction.ONE_HALF;
		}
	}
	
	public static BigFraction getOrdForYDirection(Direction d) {
		switch (d) {
		case UP:
			return BigFraction.ZERO;
		case DOWN:
			return BigFraction.ONE;
		default:
			return BigFraction.ONE_HALF;
		}
	}
	
	@Override
	public boolean hasDimension() {
		return false;
	}

	private BigFraction xOrd, yOrd;
	private List<Anchor> anchors = new ArrayList<AbstractAnchoringVertex.Anchor>(4);
	
	public List<Anchor> getAnchors() {
		return anchors;
	}

	public MultiCornerVertex(String id, BigFraction xOrd, BigFraction yOrd) {
		super(id+"_"+xOrd+"_"+yOrd);
		this.xOrd = xOrd;
		this.yOrd = yOrd;
	}
	
	public BigFraction getXOrdinal() {
		return xOrd;
	}
	
	@Override
	public void setX(double x) {
		super.setX(x);
		for (Anchor anchor : anchors) {
			anchor.setX(x);
		}
	}

	@Override
	public void setY(double y) {
		super.setY(y);
		for (Anchor anchor : anchors) {
			anchor.setY(y);
		}
	}

	public BigFraction getYOrdinal() {
		return yOrd;
	}
	
	public void addAnchor(HPos lr, VPos ud, DiagramElement underlying) {
		anchors.add(new Anchor(ud, lr, underlying));
	}
	
	public VPos getVPosFor(DiagramElement c) {
		for (Anchor anchor : anchors) {
			if (anchor.getDe() == c) {
				return anchor.getUd();
			}
		}
		
		throw new Kite9ProcessingException("No anchor found for container "+c);
	}
	
	public HPos getHPosFor(DiagramElement c) {
		for (Anchor anchor : anchors) {
			if (anchor.getDe() == c) {
				return anchor.getLr();
			}
		}
		
		throw new Kite9ProcessingException("No anchor found for container "+c);
	}
	
	public boolean hasAnchorFor(DiagramElement c) {
		for (Anchor anchor : anchors) {
			if (anchor.getDe() == c) {
				return true;
			}
		}
		
		return false;
	}
	
	@Override
	public boolean isPartOf(DiagramElement c) {
		return hasAnchorFor(c);
	}

	@Override
	public Set<DiagramElement> getDiagramElements() {
		return anchors.stream().map(a -> a.getDe()).collect(Collectors.toSet());
	}

	@Override
	public DiagramElement getOriginalUnderlying() {
		throw new Kite9ProcessingException("this is multi-element");
	}
}

File: src/main/kotlin/org/kite9/diagram/common/elements/vertex/MultiCornerVertex.kt
package org.kite9.diagram.common.elements.vertex

import org.kite9.diagram.common.fraction.LongFraction
import org.kite9.diagram.common.fraction.LongFraction.Companion.ONE
import org.kite9.diagram.common.fraction.LongFraction.Companion.ONE_HALF
import org.kite9.diagram.common.fraction.LongFraction.Companion.ZERO
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.HPos
import org.kite9.diagram.model.position.VPos

/**
 * Represents corners of diagrams, containers and any other rectangular content.
 * For elements with grid-layout, these also represent points within the grid that will need to be connected up.
 * Multi-corners can be the corners of multiple different diagram elements.
 */
class MultiCornerVertex(id: String, val xOrdinal: LongFraction, val yOrdinal: LongFraction) :
    AbstractAnchoringVertex(id + "_" + xOrdinal + "_" + yOrdinal), MultiElementVertex {

    override fun hasDimension(): Boolean {
        return false
    }

    private val anchors: MutableList<Anchor> = ArrayList(4)

    fun getAnchors(): List<Anchor> {
        return anchors
    }

    override var x: Double
        get() = super.x
        set(x) {
            super.x = x
            for (anchor in anchors) {
                anchor.setX(x)
            }
        }

    override var y: Double
        get() = super.y
        set(y) {
            super.y = y
            for (anchor in anchors) {
                anchor.setY(y)
            }
        }

    fun addAnchor(lr: HPos?, ud: VPos?, underlying: DiagramElement) {
        anchors.add(Anchor(ud, lr, underlying))
    }

    fun getVPosFor(c: DiagramElement): VPos? {
        for (anchor in anchors) {
            if (anchor.de === c) {
                return anchor.ud
            }
        }
        throw LogicException("No anchor found for container $c")
    }

    fun getHPosFor(c: DiagramElement): HPos? {
        for (anchor in anchors) {
            if (anchor.de === c) {
                return anchor.lr
            }
        }
        throw LogicException("No anchor found for container $c")
    }

    fun hasAnchorFor(c: DiagramElement): Boolean {
        for (anchor in anchors) {
            if (anchor.de === c) {
                return true
            }
        }
        return false
    }

    override fun isPartOf(c: DiagramElement?): Boolean {
        return if (c == null) false else hasAnchorFor(c)
    }

    override fun getDiagramElements(): Set<DiagramElement> {
        return anchors
            .map { a: Anchor -> a.de }
            .toSet()
    }

    companion object {

		fun isMin(b: LongFraction): Boolean {
            return b.equals(ZERO)
        }

		fun isMax(b: LongFraction): Boolean {
            return b.equals(ONE)
        }

		fun getOrdForXDirection(d: Direction?): LongFraction {
            return when (d) {
                Direction.LEFT -> ZERO
                Direction.RIGHT -> ONE
                else -> ONE_HALF
            }
        }

		fun getOrdForYDirection(d: Direction?): LongFraction {
            return when (d) {
                Direction.UP -> ZERO
                Direction.DOWN -> ONE
                else -> ONE_HALF
            }
        }
    }

}